### PR TITLE
replace --controller-debug command line argument with Qt logging categories

### PR DESCRIPTION
**Pull request by Be-ing**
Logging is separated into several subcategores:
controller.CONTROLLERNAME: general messages
controller.CONTROLLERNAME.input: incoming data
controller.CONTROLLERNAME.output: outgoing data

CONTROLLERNAME is the lowercase device name with non-alphanumeric
characters removed.

This allows more fine tuned control of what messages are shown
in the console. This is required for the logging output to be
useful with controllers that continually send input such as the
Native Instruments Traktor Kontrol S4 Mk3 and Gemini GMX (ping @codecat).

### CODE REVIEW

File: src/controllers/controller.cpp
Comment (Be-ing) [inline src/controllers/controller.cpp:31]: This is ugly but it's the only way I got the QLoggingCategory constructor to accept a `const char*`. Trying to use `qPrintable`/`QString::toLocal8Bit().constData()` does not work because the data the `const char*` points to immediately goes out of scope.

**Comment by Swiftb0y** [comment]
I can't comment much on the code without researching the Qt APIs myself first. @uklotzde is probably the more fitting person to review this. 
Conceptually, the changes seems sensible though.

**Comment by Be-ing** [comment]
Unfortunately there is no API to set the QLoggingCategory that `console.*` calls from JavaScript get sent to; they all go to the `js` logging category. I made a [feature request for Qt](https://bugreports.qt.io/browse/QTBUG-95958) to add that.

**Comment by Be-ing** [comment]
The way the filter rules work confuses me.
```
QT_LOGGING_RULES="controller*input.debug=false;controller.*.debug=true"
```
and
```
QT_LOGGING_RULES="controller.*.debug=true;controller*input.debug=false"
```
do not work to hide the input messages from the controller, but
```
QT_LOGGING_RULES="controller.*.debug=true;*.input.debug=false"
```
does work. :shrug:

File: src/controllers/controller.h
Comment (uklotzde) [inline src/controllers/controller.h:141]: How about using `.send` and `.recv` instead of `.output` and `.input`? We already use the terms *send* and *receive* in C++ code.

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (uklotzde) [inline src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: Storing a reference is not recommended. Either construct a new `QLoggingCategory` or store a shared pointer. I would prefer the first option, i.e. no indirection and pointer sharing.

Unfortunately, `QLoggingCategory` is not copyable.

File: src/controllers/controller.cpp
Comment (uklotzde) [reply src/controllers/controller.cpp:31]: `QLoggingCategory` stores the passed `const char *` as a member  :scream: 

https://doc.qt.io/qt-5/qloggingcategory.html#QLoggingCategory-1

> Note: category must be kept valid during the lifetime of this object.

File: src/controllers/scripting/controllerscriptenginebase.h
Comment (uklotzde) [inline src/controllers/scripting/controllerscriptenginebase.h:50]: Once again, storing references is unsafe without any control of their lifetimes.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: I presumed that it does not work to create multiple QLoggingCategories that share the same category string because it is not copyable, but I have not actually tried. std::shared_ptr could work though.

File: src/controllers/controller.cpp
Comment (Be-ing) [reply src/controllers/controller.cpp:31]: Yeah, I have no idea why the QLoggingCategory constructor doesn't just copy a QString...

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: I'll use a shared pointer to avoid having to pass a `const char*` around...

File: src/controllers/controller.h
Comment (Be-ing) [reply src/controllers/controller.h:141]: I don't know why but `input` and `output` feel like better terms for this... @Swiftb0y what do you think?

Comment (uklotzde) [reply src/controllers/controller.h:141]: Ok for me.

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (uklotzde) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: Idea: Create a copyable wrapper around QLoggingCategory that internally stores both the category as a QByteArray and the QLoggingCategory as a shared_pointer. It should provide an implicit conversion operator to `const QLoggingCategory&`.

@daschuer Or are instances of QLoggingCategory supposed to be created *on-the-fly* instead of copying them?

Comment (uklotzde) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: The wrapper could be named `DynamicLoggingCategory`.

The shared pointer is probably not needed. You simply instantiate a new QLoggingCategory instance from the category in the QByteArray member.

Comment (uklotzde) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: My understanding is that QLoggingCategory represents just an entry in the logger registry. Multiple instances share the same entry -> *Flyweight pattern*. Unfortunately, this is not documented properly.

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: > I presumed that it does not work to create multiple QLoggingCategories that share the same category string because it is not copyable

I was wrong. That does work: https://github.com/Be-ing/mixxx/commit/50f659cecb55c68df66e0037dda694500be2015d

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: I think it could work to pass `QByteArray(m_logBaseRawString.constData())` between classes, but then classes that need to pass the string onto another class they create would need both a QByteArray and QLoggingCategory member... ugly... :/

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: Oh, [QLoggingCategory::categoryName](https://doc.qt.io/qt-5/qloggingcategory.html#categoryName) returns the `const char*`, so no need to store a QByteArray in each class. I think using `QByteArray(m_logBase.categoryName())` to make a deep copy of the string to pass between classes would work.

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: I sure wish specifically defined lifetimes checked by the compiler were part of the language right now... :crab:

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: Alternatively we could just pass the `const char*` around directly without making copies. The lifetimes of all the objects which passed the `const char*` are within the lifetimes of the Controller so that would work, but it could be dangerous with future refactoring. I don't have a strong opinion, what do you think?

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: Or we could pass a plain `QByteArray` around and rely on Qt's implicit sharing

Comment (uklotzde) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:91]: No hacks, please. Use a wrapper with an internal QByteArray for dynamic logging categories.

**Comment by uklotzde** [comment]
> The way the filter rules work confuses me.
> 
> ```
> QT_LOGGING_RULES="controller*input.debug=false;controller.*.debug=true"
> ```
> 
> and
> 
> ```
> QT_LOGGING_RULES="controller.*.debug=true;controller*input.debug=false"
> ```
> 
> do not work to hide the input messages from the controller, but
> 
> ```
> QT_LOGGING_RULES="controller.*.debug=true;*.input.debug=false"
> ```
> 
> does work. shrug

Probably similar to GitHub actions expression syntax where there simple wildcard `*` is not recursive and does not match separators. Unfortunately, there doesn't seem to exist a recursive `**` wildcard for Qt logging rules.

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h]: I implemented a wrapper. The convenience comes at the cost of [grabbing a mutex](https://code.woboq.org/qt5/qtbase/src/corelib/io/qloggingregistry.cpp.html#_ZN16QLoggingRegistry16registerCategoryEP16QLoggingCategory9QtMsgType) with each implicit conversion to to QLoggingCategory. I think that cost is worth it considering logging is too slow for real time code anyway.

File: src/util/dynamicloggingcategory.h
Comment (Be-ing) [inline src/util/dynamicloggingcategory.h:22]: The documentation is longer than the implementation :grimacing: I wish QLoggingCategory just took care of this for us.

Comment (uklotzde) [reply src/util/dynamicloggingcategory.h:23]: They probably adopted QLoggingCategory straight from their embedded or automotive divisions where everything is static. Its design doesn't really fit for more general use cases.

File: src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
Comment (uklotzde) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h]: Ouch, I should have looked into the implementation of QLoggingCategory. Then using a `smart_ptr` is probably better.

We need a `SharedLoggingCategory` that is both copyable and movable. Internally it stores a `smart_ptr` to an immutable, non-copyable/non-movable `DynamicLoggingCategory` (all members const, copy/move ctor deleted) which contains both the category string as a `QByteArray` and the actual `QLoggingCategory`.

Could this work?

Comment (Be-ing) [reply src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h]: I found a simpler solution. I gave DynamicLoggingCategory a QLoggingCategory member and made the `operator()()` return a const reference to the member instead of creating a new QLoggingCategory for each call. I also realized I could make the members private and the copy constructor can still access the QByteArray.

File: src/util/dynamicloggingcategory.h
Comment (Be-ing) [reply src/util/dynamicloggingcategory.h:25]: I'll report a bug on the Qt bug tracker about this clumsy API after we merge this.

**Comment by Be-ing** [comment]
The pre-commit failure is complaining about a legacy mapping script. Fixing that is beyond the scope of this PR. We cannot simply apply eslint's patches because [they can break scripts](https://github.com/mixxxdj/mixxx/pull/3042#issuecomment-717241615).

**Comment by Be-ing** [comment]
I renamed DynamicLoggingCategory to RuntimeLoggingCategory. I think DynamicLoggingCategory is a bit misleading because it could be misunderstood to mean that the category name can be changed after the class is instantiated.

File: src/controllers/midi/hss1394controller.cpp
Comment (uklotzde) [inline src/controllers/midi/hss1394controller.cpp:63]: The format string is redundant.

File: src/controllers/midi/portmidicontroller.cpp
Comment (uklotzde) [inline src/controllers/midi/portmidicontroller.cpp:12]: Use QString::fromLocal8Bit() or whatever is suitable for the conversion

File: src/util/runtimeloggingcategory.h
Comment (uklotzde) [inline src/util/runtimeloggingcategory.h:46]: - Coding style: `m_` prefix is missing
- Replicating the type name is not helpful, use `m_category` instead

File: src/util/runtimeloggingcategory.h
Comment (uklotzde) [inline src/util/runtimeloggingcategory.h:28]: - `explicit`
- `oldByteArray` -> `category`
- Pass QByteArray by value and use `std::move` for initializing the member

File: src/util/runtimeloggingcategory.h
Comment (uklotzde) [inline src/util/runtimeloggingcategory.h:34]: - `oldCategory` -> `other` or `that`
- Use a delegating ctor: `RuntimeLoggingCategory(other.m_byteArray)`

**Review by uklotzde — CHANGES_REQUESTED** [review]
Please check for all trivial format strings like `"%1"` that should be replaced.

File: src/util/runtimeloggingcategory.h
Comment (uklotzde) [inline src/util/runtimeloggingcategory.h]: I would prefer to make the expensive *cloning* operation more explicit instead of hiding it behind an innocent copy ctor.

Is this possible without any copy/move ctor?

```
static RuntimeLoggingCategory clone(const RuntimeLoggingCategory& other) {
    return RuntimeLoggingCategory(other.m_category);
}
```

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: How is this expensive? [QByteArray's copy constructor](https://doc.qt.io/qt-5/qbytearray.html#QByteArray) uses implicit sharing.

Comment (uklotzde) [reply src/util/runtimeloggingcategory.h]: It's about semantics, not performance. Pass by value and move into place for taking ownership.

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: Like [this](https://github.com/mixxxdj/mixxx/pull/4239/commits/d63269da2dfb73429194e70bc1ee9cf3e274a774)? I don't really get the point, but okay.

Comment (uklotzde) [reply src/util/runtimeloggingcategory.h]: If you pass-by-value the caller can decide to either give up ownership by using std::move on their side or by passing in a copy and retaining ownership. If you pass-by-reference the copy ctor will be invoked regardless of the intention of the caller, because you **have** to copy it.

Comment (uklotzde) [reply src/util/runtimeloggingcategory.h]: Just a convention. If you only need to *borrow* temporarily for the duration of the call use pass-by-const-ref otherwise use pass-by-value to take take ownership.

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: I don't understand. If I change the constructors of the classes that take a RuntimeLoggingCategory to take a `const RuntimeLoggingCategory` instead of a `const RuntimeLoggingCategory&`, initializing the `const RuntimeLoggingCategory` members fails to compile because the copy constructor is private.

Comment (uklotzde) [reply src/util/runtimeloggingcategory.h]: Unfortunately, the Core Guidelines now recommend to pass via `&&` for consuming parameters, which is less convenient for the caller:

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-consume

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: Changing the parameter type to `const RuntimeLoggingCategory&&` still tries to invoke the copy constructor which is private.

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: ... unless you want the class members to be `const RuntimeLoggingCategory&&` too instead of `const RuntimeLoggingCategory`?

Comment (uklotzde) [reply src/util/runtimeloggingcategory.h]: If it is not possible in C++ then let's forget about it.

Alternative option: Define a custom constructor that takes an empty marker `struct Clone {}` as 2nd unused parameter and otherwise works exactly as the current copy constructor. That could not be invoked accidentally.

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: > let's forget about it

I agree. I just removed the last commit making the copy constructor private. Let's not try too hard to bend C++ into something that it is not.

File: src/controllers/midi/midioutputhandler.cpp
Comment (uklotzde) [inline src/controllers/midi/midioutputhandler.cpp:17]: Comment on wrong line. Delete it. Should be replaced by a constant eventually.

File: src/controllers/controller.cpp
Comment (uklotzde) [inline src/controllers/controller.cpp:15]: `loggingCategoryPrefix(...)`

File: src/controllers/controller.cpp
Comment (uklotzde) [inline src/controllers/controller.cpp:13]: This constant along with some utility functions belongs to RuntimeLoggingCategory which currently accepts arbitrary strings without any validation.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/controllers/controller.cpp
Comment (uklotzde) [reply src/controllers/controller.cpp:13]: ...or are these restrictions specific for this particular use case and in general arbitrary strings are allowed?

Comment (Be-ing) [reply src/controllers/controller.cpp:13]: good idea

Comment (Be-ing) [reply src/controllers/controller.cpp:13]: The restrictions are not specific to this use case:

> Category names are free text; to configure categories using Logging Rules, their names should follow this convention:
> Use letters and numbers only.
> Use dots to further structure categories into common areas.
> Avoid the category names: debug, info, warning, and critical.
> Category names with the qt prefix are solely reserved for Qt modules.

https://doc.qt.io/qt-5/qloggingcategory.html#creating-category-objects

File: src/controllers/midi/midioutputhandler.cpp
Comment (Holzhaus) [reply src/controllers/midi/midioutputhandler.cpp]: And please don't name the constant "virgin"...

Comment (Be-ing) [reply src/controllers/midi/midioutputhandler.cpp]: Don't worry, this isn't Emacs ruled by a self-appointed dictator for life without any more constructive things to say. I changed it to actually describe what this constant is https://github.com/mixxxdj/mixxx/pull/4239/commits/50c2ab588cbd0eaaa48356117072e3f5b8e5a903

File: src/util/runtimeloggingcategory.h
Comment (uklotzde) [inline src/util/runtimeloggingcategory.h:45]: - `sanitizeCategoryName()` would be less redundant and also matches the naming of the member
- Do we still use the verb *sanitize*? Other options?

Comment (Be-ing) [reply src/util/runtimeloggingcategory.h]: "Sanitize" is one of those common programming terms that isn't very self-explanatory so yes I think it is good to rename it. I renamed it to `removeInvalidCharsFromCategory`.

**Comment by uklotzde** [comment]
The `controller-debug` argument is still present even though the commit message states "replace --controller-debug CLI argument with Qt logging categories"!?

**Comment by Be-ing** [comment]
Oh, good point. I'll remove it from the `--help` message.

**Comment by uklotzde** [comment]
Please also search for `--controllerDebug` there are more occurrences.

**Comment by Be-ing** [comment]
Manual PR: https://github.com/mixxxdj/manual/pull/424

**Review by daschuer — CHANGES_REQUESTED** [review]
When this is merged, our mixxx.log file will be either spammed with midi messages or does not contain the low frequent debug messages about controllers. I think it is required to tweak the system in a way that the amount of logged data is kept unchanged by default. 

With this PR we get a new feature that allows easy filtering of logging messages, unfortunately it is much harder to use compared to our simple --controllerDebug. 

After merging this Mixxx will output 
```Unknown option controllerDebug``` 

This happens for instance when when one follows an old hint from the forums. 

The old option was easy to discover with ```mixxx --help``` the new method is not discoveable, if you don't know what you looking for. 

All this that's bad and needs to be improved. 

We may consider to print a help message that describes the equivalent. 

Please also remember that starting an app with an environment variable is WAY harder than on Linux. 
 
Did you consider to set the equivalent rules via `---controllerDebug` in case no other QT_LOGGING_RULES are set?

**Comment by Be-ing** [comment]
> When this is merged, our mixxx.log file will be either spammed with midi messages or does not contain the low frequent debug messages about controllers. I think it is required to tweak the system in a way that the amount of logged data is kept unchanged by default.

The solution to this is to fix the logging in the rest of Mixxx to use Qt's logging levels appropriately. We have to start somewhere unless you want an enormous pull request that will accumulate tons of merge conflicts. Currently almost everything goes to debug which is not how it is supposed to work. Most of that should go to the info log level. This PR fixes the controller system so log messages go to the appropriate info/debug/warning category whereas everything went to debug before.

> The old option was easy to discover with mixxx --help the new method is not discoveable, if you don't know what you looking for.

I am in doubt many people have found `--controllerDebug` from the `--help` message. I do not know why someone would expect `--controllerDebug` to exist if they had not read the controller mapping documentation which we can easily update.

> Did you consider to set the equivalent rules via ---controllerDebug in case no other QT_LOGGING_RULES are set?

That would be technically possible but I think it's extra complexity we don't need.

**Comment by daschuer** [comment]
> The solution to this is to fix the logging in the rest of Mixxx to use Qt's logging levels appropriately. 

Do you plan to hold this back until than? If not we need a solution here. 

> That would be technically possible but I think it's extra complexity we don't need. 

I think it is needed to provide a smooth transition. IMHO can't remove a simple to use feature without a single hint and replace it with a complex feature without a single hint. I also think this feature is way to complex to use at least on windows.

**Comment by Be-ing** [comment]
I am working on a hack to avoid writing the controller I/O to the log file but still show it in the console.

**Comment by daschuer** [comment]
Maybe you could outline you ideas before writing?

Why did you consider your solution a hack? I think we have use cases for writing the midi messages to the log and not.

**Comment by Be-ing** [comment]
> Why did you consider your solution a hack?

It is a hack around the improper use of logging levels and lack of categorized logging in the rest of Mixxx.

> I think we have use cases for writing the midi messages to the log and not.

I think it's fine to simply skip writing controller I/O to the log. I cannot recall ever looking at a user's (or my own) log for the controller input messages to figure something out.

**Comment by Be-ing** [comment]
> I also think this feature is way to complex to use at least on windows.

This is an issue of documentation. It's not significantly harder to set an environment variable than it already is to run Mixxx from a command prompt.

**Comment by daschuer** [comment]
> It's not significantly harder to set an environment variable

Not for you. But believe me it is! Especially if you have never heard of environment variables or if you trying to follow a Linux example on windows. 

You can easily edit your desktop shortcut to add a command line option. This is not possible for environment variables. 
See this for instance: 
https://netlicensing.io/blog/2012/06/15/set-environment-variables-in-windows-shortcut/

**Comment by Be-ing** [comment]
> You can easily edit your desktop shortcut to add a command line option.

How is that relevant if you have to run Mixxx from a command prompt anyway to see the output?

**Comment by Be-ing** [comment]
Whoops, I broke all debug logging in my last commit. I fixed it now. Controller I/O is not written to the log file by default. Whether it is written to the log file is now controlled by the `--log-level debug` command line argument instead of `--controllerDebug`.

**Comment by Be-ing** [comment]
Long term I think we should get rid of the `--log-level` command line argument and just use Qt logging filters.

**Comment by uklotzde** [comment]
> > It's not significantly harder to set an environment variable
> 
> Not for you. But believe me it is! Especially if you have never heard of environment variables or if you trying to follow a Linux example on windows.

 I disagree. This is a debugging option for developers, not for regular users.

**Comment by uklotzde** [comment]
> I think it's fine to simply skip writing controller I/O to the log. I cannot recall ever looking at a user's (or my own) log for the controller input messages to figure something out.

I agree. If it is possible to exclude all controller debug messages form the log file then let's do this.

**Comment by Be-ing** [comment]
Currently controller I/O is logged only when `--log-level debug` is used. It could easily be changed to always exclude controller I/O from the log file. I don't really care either way.

**Comment by uklotzde** [comment]
Keeping `--controller-debug` and using it to programmatically enable all controller logging categories without the need for environment variables is an acceptable compromise to maintain backward compatibility. It would be convenient as a starting point.

@Be-ing Is this possible with reasonable effort?

**Comment by Be-ing** [comment]
> Is this possible with reasonable effort?

Maybe. It might bring back https://bugs.launchpad.net/mixxx/+bug/1871238 ... which might not be an issue when we switch macOS to vcpkg with static linking.

**Review by daschuer — CHANGES_REQUESTED** [review]
I think we should keep `--controller-debug` and `--controllerDebug` working by setting up the logging category programmatically as proposed before merge. I don't see a reason that this can bring back any bug, because it will only do what will happen if the environment variable is set.

**Comment by Be-ing** [comment]
https://github.com/mixxxdj/mixxx/blob/main/src/util/logging.cpp#L381
I'm not touching that.

**Comment by daschuer** [comment]
It looks like https://github.com/mixxxdj/mixxx/pull/2635 was a not working attempt to fix https://bugs.launchpad.net/mixxx/+bug/1871238 we have missed to revert. It was merged on 2020-08-04 to git7232 and we have a report that git7243 is still crashing. It was finally fixed on 2020-04-12 by https://github.com/mixxxdj/buildserver/pull/86.

**Comment by Be-ing** [comment]
Thank you for digging into that history. We should remove the misleading comment.

**Comment by Be-ing** [comment]
I think manipulating the filter rules with the legacy `--controller-debug` option is not a good idea. Now the controller debug messages are shown with `--log-level debug`. There is no need for an extra command line switch to enable the controller debug messages separately. If you want to disable the controller debug messages and only see the other debug messages, you can set the Qt logging rules.

**Comment by daschuer** [comment]
I don't really mind, but the current state was just fine IMHO. Enable controller debug by default will puts the burden on everyone to find out how to get rid of the fast scrolling debug output when dealing with an other unrelated problem. From other projects I know that debug categories have to be enabled one by one on demand.

What do others think? 

The alternative would be to print a message how to switch controller debugging when someone uses --controlerDebug

**Comment by Be-ing** [comment]
Merge?

**Comment by daschuer** [comment]
> Merge?

No, it is a regression to fail if the --controlerDebug flag is set without enable or suggest the workaround.

I would prefer to keep the old default behavior unchanged. This way we can also remove the log file hack introduced here.

**Comment by Be-ing** [comment]
We do need to document new behavior, but I don't think we need to add an explanation in Mixxx for every behavior change, especially not for something that only affects developers.

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:263]: - Add separators to search strings, i.e. `controller.`, `.input`, and `.output`?
- Move code into `inline bool isControllerDebugLoggingCategory(const QString& categoryName)`
- Only evaluate this function which involves a string search if `WriteFlag::File` is not already set

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/util/logging.cpp
Comment (Be-ing) [reply src/util/logging.cpp:263]: > Add separators to search strings, i.e. controller., .input, and .output?

The code already does this.

Comment (Be-ing) [reply src/util/logging.cpp]: done

**Comment by Be-ing** [comment]
It is silly that this isn't merged already. If it isn't merged soon I'm just going to close it.

**Comment by daschuer** [comment]
> Be-ing force-pushed the ...

Please remember, we have the rule to not to rebase a PR under review. If a re-base is necessary please ask before.

**Comment by daschuer** [comment]
For my understanding we have exchanged strong arguments towards not silently removing "controllerDebug". 
I feel not comfortable to merge this without.

**Comment by Be-ing** [comment]
I've already responded to those arguments. I'm not going to continue arguing.

**Comment by Be-ing** [comment]
It is a shame that solving real problems continually gets obstructed by bikeshedding minor hypothetical issues.

**Comment by uklotzde** [comment]
> It is a shame that solving real problems continually gets obstructed by bikeshedding minor hypothetical issues.

With this attitude development of Mixxx will stall and get stuck.

**Comment by Swiftb0y** [comment]
I'd like to revive this PR. 
Mixxx is not a console application. We should be allowed to have breaking changes in commandline flags, especially in minor releases. If somebody wants to make use of mixxx's more advanced mapping capabilities (outside of the point-and-click mapping wizard), they already have to learn JS, read 5 long wiki pages, etc. The little effort it takes to learn how to change an applications environment variables is minuscule compared to that. 

If we want to find a middle ground, I suggest we add a couple lines to the commandline parser that catches the `--controllerDebug` flag and then prints something along the lines `controllerDebug is retired, start mixxx with this environmentvariable instead: QT_LOGGING_RULES="controller.*.debug=true;"`

**Comment by Be-ing** [comment]
This PR is still ready to merge as-is.

**Comment by Holzhaus** [comment]
I don't get really why it's such a problem to use [`QLoggingCategory::setFilterRules`] (https://doc.qt.io/qt-5/qloggingcategory.html#setFilterRules) to enable controller debug output when that flag is given (for a grace period, e. g. until 2.5). This would make life easier for a lot of users that are not computer experts.

But for me, it's also okay to merge without that. This can easily be implemented later.

**Comment by Be-ing** [comment]
I don't get why we should add a hack for hypothetical developers who could easily use the new way.

**Comment by Be-ing** [comment]
Because the rest of Mixxx doesn't use categorized logging yet, a `--controller-debug` hack can't work the way it did before. It would only be able to turn on controller debug logging without `--log-level debug`, but `--log-level debug` now turns on controller debug logging if you don't set the Qt filter rules.

**Pull request closed without merging**
